### PR TITLE
refactor: use TLS library v4

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Use the TLS Certificates library v4. Before merging this, those PR's in the lib need to be merged:
- https://github.com/canonical/tls-certificates-interface/pull/220

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I Have bumped the version of the library

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -284,10 +312,19 @@ def test_given_generic_config_is_not_valid_when_certificate_creation_request_the
...
+    )
+    @patch(
+        f"{TLS_LIB_PATH}.TLSCertificatesProvidesV4.get_certificate_requests",
     )
     def test_given_valid_config_and_pending_requests_when_update_status_then_status_is_active(  # noqa: E501
-        self, mock_set_relation_certificate
+        self,
+        patch_get_certificate_requests,
+        patch_get_outstanding_certificate_requests,
+        mock_set_relation_certificate,
```
Comment (saltiyazan) [inline tests/unit/test_charm.py:327]: Is this used?

File: tests/unit/test_charm.py
Comment (saltiyazan) [inline tests/unit/test_charm.py:1]: `add_csr_to_remote_unit_relation_data` seems not to have any useful effect in any of the tests.
But I know it's not the scope of this PR, but if you feel like it just means deleting the function and its calls :D

**Review by saltiyazan — APPROVED** [review]
Looks good, I left a few minor comments that don't block merging this PR.

File: lib/charms/lego_base_k8s/v0/lego_client.py [hunk 2]
```diff
@@ -127,13 +127,12 @@ def __init__(self, *args, plugin: str):
...
         self._logging = LogForwarder(self, relation_name="logging")
-        self.tls_certificates = TLSCertificatesProvidesV3(self, CERTIFICATES_RELATION_NAME)
-        self.framework.observe(
-            self.tls_certificates.on.certificate_creation_request,
-            self._on_certificate_creation_request,
-        )
+        self.tls_certificates = TLSCertificatesProvidesV4(self, CERTIFICATES_RELATION_NAME)
         self.cert_transfer = CertificateTransferProvides(self, CA_TRANSFER_RELATION_NAME)
         self.framework.observe(self.on.send_ca_cert_relation_joined, self._configure)
+        self.framework.observe(
```
[the remark this answers is not in the document]
Comment (kayra1) [reply lib/charms/lego_base_k8s/v0/lego_client.py:133]: Although it's good to keep this in mind, the charm is executed serially, and the observer that comes second will always run second.

**Review by kayra1 — APPROVED** [review]
LGTM

File: lib/charms/lego_base_k8s/v0/lego_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/lego_base_k8s/v0/lego_client.py:133]: This is not really a change, both the requirer and provider of most interfaces are handling relation changed events. This used to be done in the lib, now it is done in the charm but the effect is the same.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (gruyaume) [reply tests/unit/test_charm.py]: removed

File: tests/unit/test_charm.py
Comment (gruyaume) [reply tests/unit/test_charm.py:1]: Done, removed

File: lib/charms/lego_base_k8s/v0/lego_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (saltiyazan) [reply lib/charms/lego_base_k8s/v0/lego_client.py:133]: It's fine that both requirer and provider handle the event.
I meant the provider (or the requirer) handling the event twice, once in the lib and once in the charm.
But I guess it's fine based on what Kayra said.

**Pull request merged**